# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель исследования:** 
Проанализировать базу данных, ответить на вопросы.

**Ход исследования:**
- обзор данных
- анализ данных (ответы на вопросы)
- итоги исследования

Данные представлены в таблицах: `books`, `authors`, `publishers`, `ratings`, `reviews`


<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Анализ-данных-(ответы-на-вопросы)" data-toc-modified-id="Анализ-данных-(ответы-на-вопросы)-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Анализ данных (ответы на вопросы)</a></span><ul class="toc-item"><li><span><a href="#Задание-№-1" data-toc-modified-id="Задание-№-1-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Задание № 1</a></span></li><li><span><a href="#Задание-№-2" data-toc-modified-id="Задание-№-2-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Задание № 2</a></span></li><li><span><a href="#Задание-№-3" data-toc-modified-id="Задание-№-3-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Задание № 3</a></span></li><li><span><a href="#Задание-№-4" data-toc-modified-id="Задание-№-4-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Задание № 4</a></span></li><li><span><a href="#Задание-№-5" data-toc-modified-id="Задание-№-5-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Задание № 5</a></span></li></ul></li><li><span><a href="#Итоги-исследования" data-toc-modified-id="Итоги-исследования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Итоги исследования</a></span></li></ul></div>

## Обзор данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# функция для выполнения запроса

def read(query):
    res = pd.io.sql.read_sql(query, con = engine)
    return res

In [4]:
# cоздаем функцию для обзора данных

def info_table(table):
    display(table.head(10))
    print(table.info())

In [5]:
# загружаем и проводим обзор таблицы books

query_books = '''SELECT *
                 FROM books
                 '''
books = read(query_books)
info_table(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None


В таблице **`books`** представлены данные о книгах крупного сервиса для чтения книг. Таблица содержит 6 столбцов, 1000 строк, тип данных представленный в таблице object, int64.

Согласно документации к таблице:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Пропущенные значения отсутствуют, названия и типы столбцов замены не требуют.

In [6]:
# загружаем и проводим обзор таблицы authors

query_authors = '''SELECT *
                   FROM authors
                   '''
authors = read(query_authors)
info_table(authors)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None


В таблице **`authors`** представлены данные об авторах книг. Таблица содержит 2 столбца, 636 строк, тип данных представленный в таблице object, int64.

Согласно документации к таблице:

- `author_id` — идентификатор автора;
- `author` — имя автора.

Пропущенные значения отсутствуют, названия и типы столбцов замены не требуют.

In [7]:
# загружаем и проводим обзор таблицы publishers

query_publishers = '''SELECT *
                      FROM publishers
                      '''
publishers = read(query_publishers)
info_table(publishers)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None


В таблице **`publishers`** представлены данные об издательствах. Таблица содержит 2 столбца, 340 строк, тип данных представленный в таблице object, int64.

Согласно документации к таблице:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

Пропущенные значения отсутствуют, названия и типы столбцов замены не требуют.

In [8]:
# загружаем и проводим обзор таблицы ratings

query_ratings = '''SELECT *
                   FROM ratings
                   '''
ratings = read(query_ratings)
info_table(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None


В таблице **`ratings`** представлены данные о пользовательских оценках книг. Таблица содержит 4 столбца, 6456 строк, тип данных представленный в таблице object, int64.

Согласно документации к таблице:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Пропущенные значения отсутствуют, названия и типы столбцов замены не требуют.

In [9]:
# загружаем и проводим обзор таблицы reviews

query_reviews = '''SELECT *
                   FROM reviews
                   '''
reviews = read(query_reviews)
info_table(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None


В таблице **`reviews`** представлены данные о пользовательских обзорах на книги. Таблица содержит 4 столбца, 2793 строки, тип данных представленный в таблице object, int64.

Согласно документации к таблице:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

Пропущенные значения отсутствуют, названия и типы столбцов замены не требуют.

**Выводы:** Проанализировали базу данных крупного сервиса для чтения книг по подписке. База данных содержит 5 таблиц: `books`, `authors`, `publishers`, `ratings`, `reviews`. В таблице `books` представлена информация о книгах id, автор, название, количество страниц, дата публикации, издательство. Таблица `authors` содержит id и имя авторов книг. Таблица `publishers` содержит id и название издательств. Таблица `ratings` содержит данные о пользовательских оценках книг. Таблица `reviews` содержит данные о пользовательских обзорах на книги. Во всех таблицах пропущенные значения отсутствуют, названия и типы столбцов замены не требуют. Предобработка данных не требуется. 

## Анализ данных (ответы на вопросы)

###  Задание № 1
Посчитайте, сколько книг вышло после 1 января 2000 года

In [10]:
query_1 = '''SELECT COUNT(*)
             FROM books
             WHERE CAST (publication_date AS date) > '2000-01-01';
             '''

In [11]:
read(query_1)

count
0    819

**Выводы:** 
После  1 января 2000 года вышло 819 книг.

### Задание № 2
Для каждой книги посчитайте количество обзоров и среднюю оценку

In [12]:
query_2 = '''SELECT books.book_id, COUNT(DISTINCT re.review_id), AVG(ra.rating) 
             FROM books
             LEFT JOIN reviews AS re ON books.book_id = re.book_id
             LEFT JOIN ratings AS ra ON books.book_id = ra.book_id
             GROUP BY books.book_id;
             '''

read(query_2)

book_id  count       avg
0          1      2  3.666667
1          2      1  2.500000
2          3      3  4.666667
3          4      2  4.500000
4          5      4  4.000000
..       ...    ...       ...
995      996      3  3.666667
996      997      3  3.400000
997      998      4  3.200000
998      999      2  4.500000
999     1000      4  3.833333

[1000 rows x 3 columns]

**Выводы:**
В таблицах представлены данные о книгах с разным количеством обзоров и различным рейтингом.

### Задание № 3
Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [13]:
query_3 = '''SELECT  pu.publisher, COUNT(*)
             FROM books 
             LEFT JOIN publishers AS pu ON books.publisher_id = pu.publisher_id
             WHERE num_pages > 50
             GROUP BY books.publisher_id, pu.publisher
             ORDER BY COUNT(*) DESC
             LIMIT 1;
             '''
read(query_3)

publisher  count
0  Penguin Books     42

**Выводы:**
Больше всего книг (42 книги) толще 50 страниц выпустило издательство `Penguin Books`.

### Задание № 4
Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [14]:
query_4 = '''SELECT book_rat.author, AVG(book_rat.avg_rating) 
             FROM (SELECT books.book_id, COUNT(ra.rating), AVG(ra.rating) AS avg_rating, au.author
                   FROM books
                   LEFT JOIN ratings AS ra ON books.book_id = ra.book_id
                   LEFT JOIN authors AS au ON books.author_id = au.author_id
                   GROUP BY books.book_id, au.author_id
                   HAVING COUNT(ra.rating) >= 50) AS book_rat
             GROUP BY book_rat.author
             ORDER BY AVG(book_rat.avg_rating) DESC
             LIMIT 1;
             '''
read(query_4)

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

**Выводы:**
Определили автора `J.K. Rowling/Mary GrandPré` с самой высокой средней оценкой книг (4.28), для книг с 50 и более оценками.

### Задание № 5
Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [15]:
query = '''SELECT AVG(rev_count)
            FROM (SELECT username, COUNT(review_id) AS rev_count
                  FROM reviews
                  WHERE username IN (SELECT username
                                     FROM ratings
                                     GROUP BY username
                                     HAVING COUNT(rating_id) > 50)
                  GROUP BY username) AS rev_count
                 '''
read(query)

avg
0  24.333333

**Выводы:**
Среднее количество обзоров от пользователей, поставивших более 50 оценок равно 24.

## Итоги исследования

Проанализировали базу данных крупного сервиса для чтения книг по подписке. База данных содержит 5 таблиц: `books`, `authors`, `publishers`, `ratings`, `reviews`.


В таблице `books` представлена информация о книгах id, автор, название, количество страниц, дата публикации, издательство.
Таблица `authors` содержит id и имя авторов книг.
Таблица `publishers` содержит id и название издательств.
Таблица `ratings` содержит данные о пользовательских оценках книг.
Таблица `reviews` содержит данные о пользовательских обзорах на книги.

Проанализировали данные помогут для формулировки ценностного предложения для нового продукта.

После 1 января 2000 года вышло 819 книг. Большинство книг сервиса являются современными, а также имеют большое количество обзоров и имеют различный рейтинг.
Больше всего книг (42 книги) толще 50 страниц выпустило издательство Penguin Books. Самой высокой средней оценкой книг (4.28), для книг с 50 и более оценками обладает автора J.K. Rowling/Mary GrandPré. Среднее количество обзоров от пользователей, поставивших более 50 оценок равно 24. Большое количество оценок и написанных рецензий может гарантировать пользователям приложения - честные пользовательские рейтинги.
 
Ценностное предложение можно сформулировать например так: Сервис для чтения книг, которому можно доверять. Самые актуальные книги от современных авторов и проверенных издательств. Честные пользовательские рейтинги и большое количество рецензий. 